url1a41b8e9f2_mr1azl_INF345_tgz-pJ8-TestParserJ8/src/xx/MonCompiler.java

```diff
@@ -0,0 +1,2234 @@
+		ret.setNodeType(NODE_TYPE_NUMBER);
+		return ret;
+	}
+
+	/**
+	 * retourne le parametre en String
+	 * @param value String
+	 * @return Object
+	 * @see xx.Compiler#literal(String)
+	 */
+	@Override
+	public Object literal(String value) {
+		NodeTest ret = new NodeTest();
+		ret.setNodeValue(Utils.asString((value)));
+		ret.setNodeType(NODE_TYPE_STRING);
+		return ret;
+	}
+
+/**
+	 * [12]   	STag	   ::=   	'<' QName (S Attribute)* S? '>'	
+	 * [13]   	ETag	   ::=   	'</' QName S? '>' 
+	 * [14]   	EmptyElemTag	   ::=   	'<' QName (S Attribute)* S? '/>'
+	 * @param prefix String
+ * @param name String
+ * @return Object
+ * @see xx.Compiler#qname(String, String)
+ */
+	@Override
+	public Object qname(String prefix, String name) {
+		return new QName(prefix, name);
+	}
+
+	/**
+	 * Si aucun argument n'est un chemin relative retourne la somme des argument
+	 * Si un des argument est un chemin relative on ajoute un appel a la
+	 * fonction sum dans la pile des fonctions
+	 * @param arguments Object[]
+	 * @return Object
+	 * @see xx.Compiler#sum(Object[])
+	 */
+	@Override
+	public Object sum(Object[] arguments) {
+		NodeTest ret = new NodeTest();
+		int val = 0;
+		NodeTest[] nodes = toNodeTestArray(arguments);
+		if (hasType(nodes, NODE_TYPE_RELATIVE_LOCATIONPATH)) {
+
+			ret.getFunctionCalls().put("sum", nodes);
+			ret.setNodeType(NODE_TYPE_RELATIVE_LOCATIONPATH);
+		}
+
+		for (int i = 0; i < arguments.length; i++) {
+//			System.out.print(getCurrentMethodName() + " ");
+			if (((NodeTest) arguments[i]).getNodeType() == NODE_TYPE_NUMBER) {
+				val += Utils.asDouble(((NodeTest) arguments[i]).getNodeValue());
+				ret.setNodeType(NODE_TYPE_NUMBER);
+				ret.setNodeValue(Utils.asString(val));
+				//System.out.println(Utils.asDouble(((NodeTest) arguments[i]).getNodeValue()));
+
+			}
+		}
+		;
+		return ret;
+	}
+
+	/**
+	 * Si aucun argument n'est un chemin relative retourne la soustraction des
+	 * argument Si un des argument est un chemin relative on ajoute un appel a
+	 * la fonction minus dans la pile des fonctions
+	 * @param left Object
+	 * @param right Object
+	 * @return Object
+	 * @see xx.Compiler#minus(Object, Object)
+	 */
+	@Override
+	public Object minus(Object left, Object right) {
+		NodeTest ret = new NodeTest();
+		double val = 0;
+
+		if (((NodeTest) left).getNodeType() == NODE_TYPE_RELATIVE_LOCATIONPATH
+				|| ((NodeTest) right).getNodeType() == NODE_TYPE_RELATIVE_LOCATIONPATH) {
+
+			ret.getFunctionCalls().put("minus",
+					new NodeTest[] { ((NodeTest) left), ((NodeTest) right) });
+			ret.setNodeType(NODE_TYPE_RELATIVE_LOCATIONPATH);
+		}
+		if (((NodeTest) left).getNodeType() == NODE_TYPE_NUMBER
+				&& ((NodeTest) right).getNodeType() == NODE_TYPE_NUMBER) {
+			val = Utils.asDouble(((NodeTest) left).getNodeValue())
+					- Utils.asDouble(((NodeTest) right).getNodeValue());
+
+			ret.setNodeType(NODE_TYPE_NUMBER);
+			ret.setNodeValue(Utils.asString(val));
+		}
+		return ret;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 145:163, 178:196, 210:227, 241:259.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.